### PR TITLE
Add compile-gcc check

### PR DESCRIPTION
**Pull request by zakhark**
#15

### CODE REVIEW

File: checks/compile_gcc.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+//
+// Support Platforms:
+//   - Mac
+//   - Linux
+//   - Windows
+//
+// Arguments:
+//   source (required): The source code of the script.
+//   interpreter: path to bash. Default is 'gcc' from the PATH
+//				  but it will also check if gcc exists in toolchain folders.
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: Need to document c_flags here

File: checks/compile_gcc.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+// Check Runs a gcc command and checks the return code
+func (cg CompileGcc) Check() error {
+
+	tmpfolder, err := ioutil.TempDir("", "compileGcc_")
+	if err != nil {
+		return fmt.Errorf("Problem creating a tmpdir: %s", err)
+	}
+
+	srcfileName := tmpfolder + "/src.c"
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: Use filepath.Join here instead of concatenation to make it cross-platform

File: checks/compile_gcc.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	outfileName := tmpfolder + "/out.o"
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: same

File: checks/compile_gcc.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+	srcfile, err := os.Create(srcfileName)
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: You don't check this error

File: checks/compile_gcc.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	defer os.RemoveAll(tmpfolder)
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: Should go right after the error checking of ioutil.TmpDir so it is close to the code it's related to.

File: checks/compile_gcc.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+	if err := decodeFromArgs(args, &cg); err != nil {
+		return nil, err
+	}
+
+	if _, interpreterGiven := args["interpreter"]; cg.Interpreter == "" && !interpreterGiven {
+
+		cg.Interpreter = "gcc"
+
+		paths := []string{
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: These paths are mongodb specific and should not be in redalert. These paths should be in our mongodb specific test file

**Comment by zakhark** [comment]
I've addressed the comments

File: checks/compile_gcc.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+//
+// Support Platforms:
+//   - Mac
+//   - Linux
+//   - Windows
+//
+// Arguments:
+//   source (required): The source code of the script.
+//   compiler: path to the compiler. Default is 'gcc' from the PATH
+//	 cflags: compiles flags, string, e.g "-lss -lsasl2"
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: Bad formatting should match indentation of the key name above it

File: checks/compile_gcc.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+	if err := decodeFromArgs(args, &cg); err != nil {
+		return nil, err
+	}
+
+	if _, compilerGiven := args["compiler"]; cg.Compiler == "" && !compilerGiven {
+		cg.Compiler = "gcc"
+	}
+
+	if _, cflagsGiven := args["cflags"]; cg.Cflags == "" && !cflagsGiven {
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: This isn't needed, the zero value for a string in Go is `""` so you don't need this if statement.

File: checks/compile_gcc.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	if runtime.GOOS == "windows" {
+		cg.Source = strings.Replace(cg.Source, "\n", "\r\n", -1)
+	}
+
+	content := []byte(cg.Source)
+
+	if _, err := srcfile.Write(content); err != nil {
+		return fmt.Errorf("Problem writing to a tmpfile: %s", err)
+	}
+	if err := srcfile.Close(); err != nil {
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: for consistency add a blank new line here

File: checks/compile_gcc.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		return fmt.Errorf("Problem closing a tmpfile: %s", err)
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: "Problem closing the tmpfile" is more grammatically correct

File: checks/compile_gcc.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+		flags := strings.TrimRight(string(out), "\n")
+		argv = append(argv, strings.Split(flags, " ")...)
+	}
+
+	cmd := exec.Command(cg.Compiler, argv...)
+	out, err := cmd.CombinedOutput()
+
+	if err != nil {
+		return fmt.Errorf("Problem running the gcc compile: %s: %s", err.Error(), string(out))
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: `err.Error()` is redundant should just be `err`

File: checks/compile_gcc.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+	if _, compilerGiven := args["compiler"]; cg.Compiler == "" && !compilerGiven {
+		cg.Compiler = "gcc"
+	}
+
+	if _, cflagsGiven := args["cflags"]; cg.Cflags == "" && !cflagsGiven {
+		cg.Cflags = ""
+	}
+
+	if _, cflagscommandGiven := args["cflags_command"]; cg.CflagsCommand == "" && !cflagscommandGiven {
```
Comment (chasinglogic) [inline checks/compile_gcc.go]: Same as above, if block not needed

**Pull request merged**